forest/db/control.py

```diff
@@ -332,23 +316,24 @@ def __call__(self, store, next_dispatch, action):
                     variable=variable,
                     initial_time=initial_time)
                 pressures = list(reversed(pressures))
-                next_dispatch(action)
-                next_dispatch(set_value("valid_times", valid_times))
-                next_dispatch(set_value("pressures", pressures))
+                yield action
+                yield set_value("valid_times", valid_times)
+                yield set_value("pressures", pressures)
                 return
             elif key == "initial_time":
                 for attr in ["pattern", "variable"]:
                     if attr not in store.state:
-                        return next_dispatch(action)
+                        yield action
+                        return
                 valid_times = self.navigator.valid_times(
                     pattern=store.state["pattern"],
                     variable=store.state["variable"],
                     initial_time=value)
                 valid_times = sorted(set(valid_times))
-                next_dispatch(action)
-                next_dispatch(set_value("valid_times", valid_times))
+                yield action
+                yield set_value("valid_times", valid_times)
                 return
```

Comment:
From a code archaeology point of view this gnarly piece of logic is my earliest implementation of middleware, like naive art, it suffers from a lack of mastery from not knowing how to correctly parse/structure actions and their consequent actions. But yet it remains unaltered as it powers the navigation system, it is in need of a spruce up.

Reply:
A better action structure instead of

```python
{"kind": KIND, "payload": {"key": "valid_time", "value": "2019-01-01 00:00:00"}}
```

Would be:

```python
{"kind": KIND, "payload": {"valid_time", "2019-01-01 00:00:00"}}
```
Then the `reducer` logic is `state['sub-component'].update(action['payload'])` and the middleware check is simpler too:

```python
if 'valid_time' in action['payload']:
    # do something
```
This opens the door to actions that set more than one property simultaneously
```python
{"kind": KIND, "payload": {
    "valid_time", "2019-01-01 00:00:00",
    "valid_times": [...]
}}
```
In enterprise message passing systems this is called a message `enhancer`, while until now we have been using a `splitter` pattern

https://www.enterpriseintegrationpatterns.com/patterns/messaging/index.html

Reply:
New issue with `tag:Refactor` perhaps?

Reply:
Changing the contents of the actions does seem like a task for another PR, but I think it would still help to refactor the existing logic in this PR.

Reply:
@rhattersley I've refactored the `__call__()` C style switch statement to delegate to `_handler` methods. I think it already makes the code more readable but I've left as an exercise for future PRs how best to clean up the handler logic and action shape